### PR TITLE
[SDL 0317] SDL Protocol Security Specification

### LINKED ISSUE #40: [SDL 0317] SDL Protocol Security Specification
Proposal: [SDL Protocol Security Specification](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0317-sdl-protocol-security-specification.md)

> This proposal adds the documentation and specification around protected communication, encryption and handshake to the SDL protocol specification. It is a collaborated effort with Ford and Luxoft to protect the implemented SDL security from unexpected changes and to allow SDL Evolution proposals to improve it.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/1070

Steering Committee Decision: The Steering Committee voted to accept this proposal with the following revisions:

1. Remove "Note: Communication partners should set this field to zero" from the "Frame Info" and "Data Size" descriptions in the Change 1 table (https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0317-sdl-protocol-security-specification.md#change-1-update-frame-header-fields-description).
2. In `Proposed solution` section (4.2.5 Start Service), change "See "Secured Communication" section for more details." to "See "7. Secured Communication" section for more details."
3. In `Proposed solution` section (7. Secured Communication), change "It is possible to establish a secured and encrypted communication with the system by setting the frame header encryption flag to 1 when starting a new service." to "It is possible to establish a secured and encrypted communication with the system by setting the frame header encryption flag to 1 when starting a new service or by sending another StartService with the encryption flag to 1 when the service is already established (this the required flow for the RPC service)."
4. Update `Impact on existing code` section to remove code changes related to Core/Mobile libraries, and update `Impacted Platforms` to only include "Protocol". 

The proposal .md file was updated to reflect these revisions on 6/10/21

### PR DESCRIPTION
**Pull request by AKalinich-Luxoft**
Implements #40

### CODE REVIEW

**Comment by AKalinich-Luxoft** [comment]
@theresalech this PR is ready for Livio review

**Comment by theresalech** [comment]
@AKalinich-Luxoft , can you please advise if this PR has already been reviewed and approved by a member of the Ford team? We will begin our review after a Ford team member has provided their approval on the PR. Thank you!

**Comment by mrapitis** [comment]
@theresalech Ford has reviewed and approved of this PR. Thanks!

File: README.md
Comment (jacobkeeler) [inline README.md]: > During the handshake the TLS handshake data is sent as binary data.

I think we should probably include a subsection `5.x.x` around here describing this, even if it just links out to the "Secured Connection" section:

```

**Request:**

<table>
  <tr>
    <th align="center">Binary Data</th>
  </tr>
  <tr>
    <td align="center" width="800">SSL Handshake Request</td>
  </tr>
</table>

**Response:**

<table>
  <tr>
    <th align="center">Binary Data</th>
  </tr>
  <tr>
    <td align="center" width="800">SSL Handshake Response</td>
  </tr>
</table>
```

This might require renumbering a few of the sections

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
      Message ID can be set by the mobile libraries to track security messages.<br>
```

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
      The system uses the same Message ID when replying to the query allowing the mobile libraries to correlate messages.
```

File: README.md
Comment (jacobkeeler) [inline README.md:1465]: These 3 values are not defined in Core, where are these coming from?

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
    <td>
      The size of the JSON data following the binary query header.<br>
      Any additional data following the JSON data in the payload is binary data.
    </td>
```

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
#### 5.1.1.2 Send Internal Error
```
This should probably be a subsection of "Security Query", as should `Send Handshake Data`

File: README.md
Comment (jacobkeeler) [inline README.md]: SSLv3 isn't supported on newer systems, I think it should either be removed or at least noted that it isn't supported on most systems.

File: README.md
Comment (jacobkeeler) [inline README.md]: This is quite a cluttered section which describes a lot of different scenarios quickly, perhaps it would be best to lay out these scenarios in table format?

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
The authentication is done using TLS handshake. The TLS handshake process is defined by TLS and is not part of the SDL protocol. The handshake is designed as a client-server communication which is configurable in the system settings. An application must take the role of a server where the system is the client. The client entity will initiate a TLS handshake with the corresponding security manager of the server. The client will do this only if the server was not authenticated before in the current transport connection. According to the TLS handshake process the peer certificate can be omitted for the server but it's required for the client. Certificate peer verification can be enabled/disabled on the Core side by changing `VerifyPeer` parameter in the configuration file. On the other hand, the SDL app library does not require the certificate from Core for the TLS handshake. However, Core performs its internal certificate validation before starting the actual TLS handshake. During internal validation, Core checks if the certificate is missing (or outdated/invalid) and if so, it initiates a PTU to obtain a new certificate from the Policy Server. If a valid certificate can't be obtained, Core does not start the TLS handshake and it notifies the app library that the protected service start has failed.
```

File: README.md
Comment (jacobkeeler) [inline README.md:1659]: I'm wondering whether the protocol spec is the best place for all of this info in `7.1`, it might be better to include this in the SDL Overview Guides and link out to it here. It has several references to app and Core configurations that probably aren't really relevant to the protocol

It might still be worth including the TLS handshake diagram here, but most of the other information I feel could be moved to a better place.

File: README.md
Comment (jacobkeeler) [inline README.md:194]: > 0x08 The data size for a first frame is always 8 bytes.

Any reason for removing this note?

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
When establishing a secure connection, the TLS Payload is sent in a control service message with a binary query header. The size of this header is 12 bytes, similar to the RPC Payload Binary Header.
```

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler moved into a separate section "Send Handshake Data"
See https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md:1465]: @jacobkeeler looks like it is the values from older SYNC generations that are not used in newer systems. I removed these three to avoid changes in SDL core. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler done. Now both "Send Handshake Data" and "Send Internal Error" are specified as separate subsections. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler I moved this one to the bottom of the list and mentioned that it is not supported on newer systems. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler I converted these 2 paragraphs into a table format. Hopefully, now it's easier to read. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md:194]: @jacobkeeler looks like it was removed by mistake. I restored the first sentence here. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

File: README.md
Comment (AKalinich-Luxoft) [reply README.md:1659]: @jacobkeeler could you advise please if [this](https://github.com/smartdevicelink/sdl_core_guides/tree/master/docs/Feature%20Documentation) is the right place for moving out the information from this section?

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
In case of an error, a notification is sent with an error code and error message as JSON data. See "Send Internal Error" section for details.
```

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
- SSLv3 (not supported on most newer systems)
```

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
<table width="100%">
  <tr>
    <th width="30%">Service Encryption Status</th>
    <th width="10%">RPC Type</th>
    <th width="10%">Requires Protection</th>
    <th width="50%">Expected SDL Behavior</th>
  </tr>
  <tr>
    <td rowspan="4">Encryption is not established</td>
    <td rowspan="2">Request</td>
    <td>yes</td>
    <td>SDL Core rejects the request with result code `ENCRYPTION_NEEDED` (please see policy updates for which RPCs need protection).</td>
  </tr> 
  <tr>
    <td>no</td>
    <td>SDL Core continues processing the RPC request.</td>
  </tr>
  <tr>
    <td rowspan="2">Notification</td>
    <td>yes</td>
    <td>SDL Core does not send the notification.</td>
  </tr> 
  <tr>
    <td>no</td>
    <td>SDL Core sends the notification unencrypted.</td>
  </tr>
  
  <tr>
    <td rowspan="4">Encryption is established</td>
    <td rowspan="2">Request</td>
    <td>yes</td>
    <td>
      If unencrypted, SDL Core rejects the request with an unencrypted response and result code `ENCRYPTION_NEEDED`.<br>
      If encrypted, SDL Core continues processing the request and sends an encrypted response.
    </td>
  </tr>
  <tr>
    <td>no</td>
    <td>
      If unencrypted, SDL Core continues processing the request and sends an unencrypted response.<br>
      If encrypted, SDL Core continues processing the request and sends an encrypted response.
    </td>
  </tr>
  <tr>
    <td rowspan="2">Notification</td>
    <td>yes</td>
    <td>SDL Core sends the notification encrypted.</td>
  </tr>
  <tr>
    <td>no</td>
    <td>SDL Core sends the notification unencrypted.</td>
  </tr>

</table>
```

Some suggestions to clean up the table

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/efddba1d5059d7aa763a2502acb600a726535001

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/efddba1d5059d7aa763a2502acb600a726535001

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler thank you, applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/2f371a70d67b8598b9cec923524479660b31024c

File: README.md
Comment (AKalinich-Luxoft) [reply README.md:1659]: @jacobkeeler updated in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/efddba1d5059d7aa763a2502acb600a726535001

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
      0x08 The data size for a first frame is always 8 bytes. In the payload, the first four bytes denote the Total Size of the data contained in all consecutive frames. This is always the size of whole non-encrypted payload (even if consecutive frames are encrypted). The second four bytes denote the number of consecutive frames following this one<br>
```

File: README.md
Comment (jacobkeeler) [reply README.md:1659]: @AKalinich-Luxoft I think the best place for this information would be the SDL Overview Guides, since it relates to both the mobile libraries and SDL Core. It might need to be in its own "Protected Services" page for now, as the existing "RPC Encryption" guide doesn't really cover encrypted video and audio services.

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/a323ff3e488b4710003d7e905dde521ac4a5e2f5

File: README.md
Comment (AKalinich-Luxoft) [reply README.md:1659]: @jacobkeeler I added a link to SDL Overview Guides in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/b25ef60451a378bf1a7adf7701eaa96481e93454
Also, I opened [PR](https://github.com/smartdevicelink/sdl_overview_guides/pull/80) containing moved changes. The link in protocol_spec will become valid once this PR is merged. Could you please review it as well?

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
Please see [SDL Overview Guides](https://smartdevicelink.com/en/guides/pull_request/sdl-overview-guides/security/protected-services/) for more details.
```
One more link fix after the latest changes to the Overview Guides PR

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/4823c337e15e4dcbf5e5918bec884f1cff192c06

**Comment by AKalinich-Luxoft** [comment]
@jacobkeeler all comments have been processed. Would you have a chance to proceed with the review here?

**Comment by jacobkeeler** [comment]
@AKalinich-Luxoft Ah, so we've been holding off on officially approving this because there is some testing we need to do to verify there aren't any legacy issues with this specification. Currently this spec isn't properly implemented in either of the mobile libraries (see https://github.com/smartdevicelink/sdl_ios/issues/2024 and https://github.com/smartdevicelink/sdl_java_suite/issues/1720), and we're in the process of fixing this. As a result, we are currently planning to test these updates against a number of legacy systems to verify that everything still works properly (and if not, we would submit a revision to the proposal which accounts for this). 

I can approve this PR for now if we need to, but we will need to hold off on merging until this testing is performed.

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
  <tr>
    <td>Notification</td>
    <td>Send Internal Error</td>
    <td>Unused</td>
    <td>Size of the JSON data</td>
  </tr>
  <tr>
    <td>0x20</td>
    <td>0x000002</td>
    <td>0xNNNNNNNN</td>
    <td>0xNNNNNNNN</td>
  </tr>
```

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
<table width="100%">
  <tr>
    <th colspan="4">Binary Query Header</th>
  <tr>
  <tr>
    <th>Query Type</th>
    <th>TLS Message Type</th>
    <th>Sequential Number</th>
    <th>JSON Size</th>
  </tr>
  <tr>
    <td>Request</td>
    <td>Send Handshake Data</td>
    <td>Any number to be used to correlate query messages</td>
    <td>Zero</td>
  </tr>
  <tr>
    <td>0x00</td>
    <td>0x000001</td>
    <td>0xNNNNNNNN</td>
    <td>0x00000000</td>
  </tr>
  <tr>
    <th align="center">Binary Data: SSL Handshake Request</td>
  </tr>
</table>
```
Should use the more detailed definitions used in the proposal for clarity

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
##### 5.1.1.2.1 Request Payload

**Note:** Prior to SDL Core version 8.0.0, Core sent the "Notification" type (`0x20`) in this message instead of "Request" (`0x00`). Client libraries should account for this when communicating with older versions of SDL Core.

```
Updating subsection names for organization and added note about previous Core behavior

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
##### 5.1.1.2.2 Response Payload

```

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
<table width="100%">
  <tr>
    <th colspan="4">Binary Query Header</th>
  <tr>
  <tr>
    <th>Query Type</th>
    <th>TLS Message Type</th>
    <th>Sequential Number</th>
    <th>JSON Size</th>
  </tr>
  <tr>
    <td>Response</td>
    <td>Send Handshake Data</td>
    <td>Any number to be used to correlate query messages</td>
    <td>Zero</td>
  </tr>
  <tr>
    <td>0x10</td>
    <td>0x000001</td>
    <td>0xNNNNNNNN</td>
    <td>0x00000000</td>
  </tr>
  <tr>
    <th align="center">Binary Data: SSL Handshake Response</td>
  </tr>
</table>
```
Should use the more detailed definitions used in the proposal for clarity

File: README.md
Comment (jacobkeeler) [inline README.md:1278]:
```suggestion
##### 5.1.1.3 Send Handshake Data

When performing the TLS handshake, the server sends a "Send Handshake Data" request containing its handshake data to the client, and the client sends a response with its own handshake data accordingly.

```
Adding description for this message.

**Review by jacobkeeler — CHANGES_REQUESTED** [review]
Added suggestions for rewriting the `Send Handshake Data` description for clarity, in addition to a note describing incorrect behavior in older versions of Core.

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler fixed in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied. I also used `5.1.1.3.1` section number as from my understanding this is a sub section of `5.1.1.3 Send Handshake Data`
See https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied and updated section numbering
See https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

File: README.md
Comment (AKalinich-Luxoft) [reply README.md:1278]: @jacobkeeler added in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
    <th align="center" colspan="4">Binary Data: SSL Handshake Response</td>
```

File: README.md
Comment (jacobkeeler) [inline README.md]:
```suggestion
    <th align="center" colspan="4">Binary Data: SSL Handshake Request</td>
```

**Review by jacobkeeler — CHANGES_REQUESTED** [review]
Missed some small formatting issues

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/26e0db45251d3f8a3667b2f66e38660467ac001c

File: README.md
Comment (AKalinich-Luxoft) [reply README.md]: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/26e0db45251d3f8a3667b2f66e38660467ac001c

File: README.md
Comment (jacobkeeler) [inline README.md:1450]:
```suggestion
  </tr>
  <tr>
    <td>ERROR_HANDSHAKE_FAILED</td>
    <td>0x09</td><td>9</td>
    <td>In case of all other handshake errors</td>
  </tr>
  <tr>
    <td>INVALID_CERT</td>
    <td>0x0A</td><td>10</td>
    <td>Handshake failed because certificate is invalid</td>
  </tr>
  <tr>
    <td>EXPIRED_CERT</td>
    <td>0x0B</td><td>11</td>
    <td>Handshake failed because certificate is expired</td>
  </tr>
```
Seems that these are included in the proposal, so they need to be in this PR, they've been implemented in Core respectively: https://github.com/smartdevicelink/sdl_core/pull/3773

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: README.md
Comment (AKalinich-Luxoft) [reply README.md:1450]: @jacobkeeler applied in 347cbb864affdcf3f4e5afd1623b845b710a5c67
Should I rename `INVALID_CERT` -> `ERROR_INVALID_CERT` and `EXPIRED_CERT` -> `ERROR_EXPIRED_CERT` like in PR?

**Review by jacobkeeler — APPROVED** [review]

**Pull request merged**
